### ISSUE TITLE
Unobserved exception was rethrown by the finalizer thread

### ISSUE THREAD

**Issue by mmaitre314** [Azure/azure-event-hubs-dotnet#283]
Our telemetry is reporting heavy exception rate coming from `AmqpConnection.AddSession()`. The error is likely expected (race condition between closing partition receivers and recreating them it seems) but it gets reported by the finalizer thread so we cannot handle them. Could those be either caught in the Event Hub library or forwarded to the caller (our code)?

- Inner Exception: `System.InvalidOperationException` `Can't create session when the connection is closing.`
- Outer Exception: `System.AggregateException` `A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. Can't create session when the connection is closing.`

Inner Exception callstack:

```
System.InvalidOperationException:
   at Microsoft.Azure.Amqp.AmqpConnection.AddSession (Microsoft.Azure.Amqp, Version=2.3.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at Microsoft.Azure.Amqp.AmqpCbsLink+OpenCbsRequestResponseLinkAsyncResult+<GetAsyncSteps>d__7.MoveNext (Microsoft.Azure.Amqp, Version=2.3.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at Microsoft.Azure.Amqp.AsyncResult.End (Microsoft.Azure.Amqp, Version=2.3.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at Microsoft.Azure.Amqp.AmqpCbsLink.EndCreateCbsLink (Microsoft.Azure.Amqp, Version=2.3.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at Microsoft.Azure.Amqp.FaultTolerantAmqpObject`1+<OnCreateAsync>d__5.MoveNext (Microsoft.Azure.Amqp, Version=2.3.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at Microsoft.Azure.Amqp.Singleton`1+<CreateValue>d__20.MoveNext (Microsoft.Azure.Amqp, Version=2.3.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
```

## Actual Behavior
1. Exception reported on finalizer thread

## Expected Behavior
1. Exception reported to caller code

## Versions  
- OS platform and version: Windows 10
- .NET Version: Microsoft.Azure.EventHubs
- NuGet package version or commit ID: 2.0.0

**Comment by serkantkaraca** [comment]
This is an issue in the AMQP library. Can you report it here? https://github.com/azure/azure-amqp

**Comment by mmaitre314** [comment]
Thanks! Moved.

**Comment by serkantkaraca** [comment]
Btw, there is another issue we are tracking which seems to be related to this one. An AMQP developer is already working on this.

**Comment by mmaitre314** [comment]
Thanks. That will be helpful: we are seeing peaks at 160K exceptions/5 minutes on some of our clusters and that hides real issues.

![image](https://user-images.githubusercontent.com/8584604/41138998-d62b3058-6a99-11e8-9881-389d9d314f3f.png)

**Comment by serkantkaraca** [comment]
What type of exceptions are those? You can distinguish service side exceptions from client side ones by looking at the tracking id in the exception message.
